Add guide on storing sensitive data (#1562)
vaadin/docs: articles/security/advanced-topics/external-configuration.adoc

---
title: Externalize Configuration Properties

> layout: page

Reviewer: ```suggestion
layout: page
order: 250
```

---
title: Externalize Configuration Properties
layout: page
order: 250
---

== Use System Environment Variables

> If your `application.properties` file contains sensitive information, such as database URL, username, or password, 

Reviewer: This appears to be an incomplete sentence.

> It is often a bad practice to put sensitive information, such as database URL, username, or password in your `application.properties` files. One of the easiest way to externalize sensitive information is to set those values as system environment variables.

Reviewer: > It is often a bad practice to put sensitive information ... in your `application.properties` files.

We should say a word or two about why is it a bad practice to do so. It is also not just about `application.properties` file: it is a bad practice to store sensitive info and git commit them anywhere in the project.

---
title: Externalize Configuration Properties
layout: page
order: 250
---

It is often a bad practice to put sensitive information, such as database URI, username, or password in your `application.properties` files.
If sensitive information is stored in project files or in code, then they might leak when you commit your project into source control management systems.

To avoid leaking sensitive information, you should consider storing sensitive information outside of your project files.

> 

Reviewer: ```suggestion

This guide demonstrates two ways to externalize sensitive data:
* Using system environment variables
* Using an external properties file

```

== Use System Environment Variables

One of the easiest way to externalize sensitive information is to set those values as system environment variables.

> For example, a Vaadin Start project configured to use PostgreSQL might contain the sensitive information below in the `application.properties` file.

Reviewer: ```suggestion
For example, a https://start.vaadin.com/[Vaadin Start] project configured to use PostgreSQL might contain the sensitive information below in the `application.properties` file.
```

.application.properties
[source,properties]
----
...
spring.datasource.url = jdbc:postgresql://localhost:5432/postgres
spring.datasource.username = postgres
spring.datasource.password = password
...
----

To replace the sensitive information with system environment variables, you edit your file like below.

.application.properties
[source,properties]
----
...
spring.datasource.url = ${DB_URL}
spring.datasource.username = ${DB_USER}
spring.datasource.password = ${DB_PASSWORD}
...
----

The `${...}` token is the special syntax for Spring's https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.external-config.files.property-placeholders[property placeholder], while `DB_URL`, `DB_USER`, and `DB_PASSWORD` represent the actual environment variables that you need to set in your own system.

> On Linux and MacOS, to temporarily set the environment variables, you can use the commands below in your IDE's terminal.

Reviewer: ```suggestion
On Linux and macOS, to temporarily set the environment variables, you can use the commands below in your IDE's terminal.
```

[source,zsh]
----
export DB_URL=jdbc:postgresql://localhost:5432/postgres
export DB_USER=postgres
export DB_PASSWORD=password
----

If you want to set the environment variables permanently, however, you must add the commands above into your `~/.zprofile` (_zsh_) or `~/.bashrc` (_bash_) files, depending on your system's default shell.

On Windows, you can set the environment variable by:

. Navigating to *Control Panel* > *System*.
. In *System Properties*, select *Advanced* > *Environment Variables*.
. Add new user or system environment variables by using the *New* buttons.

image::images/windows_env.png[Add environment variable on Windows.]

== Import Configuration From External File

Another simple solution to externalize sensitive information would be to import external properties files using Spring Boot's `spring.config.import` properties.

> Assuming that you have a file called `db.properties` to hold your database secrets,

Reviewer: ```suggestion
Assuming that you have ab external file called `db.properties` to hold your database secrets,
```
